Langchain Integration Fails with Langchain 0.0.300 and ddtrace 1.20.1

Pinning Langchain to < 0.0.300 fixes the issue, it appears that the Datadog patch for langchain needs to be updated.

Stacktrace:

```
  AttributeError: Could not find: llms
  ...
  File "ddtrace/vendor/wrapt/importer.py", line 177, in _exec_module
    notify_module_loaded(module)
  File "ddtrace/vendor/wrapt/decorators.py", line 470, in _synchronized
    return wrapped(*args, **kwargs)
  File "ddtrace/vendor/wrapt/importer.py", line 136, in notify_module_loaded
    hook(module)
  File "ddtrace/_monkey.py", line 167, in on_import
    imported_module.patch()
  File "ddtrace/contrib/langchain/patch.py", line 759, in patch
    wrap("langchain", "llms.base.BaseLLM.generate", traced_llm_generate(langchain))
  File "ddtrace/vendor/wrapt/wrappers.py", line 882, in wrap_function_wrapper
    return wrap_object(module, name, FunctionWrapper, (wrapper,))
  File "ddtrace/vendor/wrapt/wrappers.py", line 822, in wrap_object
    (parent, attribute, original) = resolve_path(module, name)
  File "ddtrace/vendor/wrapt/wrappers.py", line 808, in resolve_path
    original = lookup_attribute(parent, attribute)
  File "ddtrace/vendor/wrapt/wrappers.py", line 806, in lookup_attribute
    return getattr(parent, attribute)
  File "__init__.py", line 308, in __getattr__
    raise AttributeError(f"Could not find: {name}")
```

Error does not appear without ddtrace enabled.

I'm also having the same issue. I've checked the previous versions, and the issue started with version `0.0.302`. It only happens when use `ddtrace-run`. Looking at the commits from that version and the error, it seems to have something to do with the added `TYPE_CHECKING`.

Maybe [this](https://github.com/langchain-ai/langchain/commit/c87e9fb2ce0ae617e3b2edde52421c80adef54cc) commit is the culprit.

My full log:
```
2023-10-05T14:11:39.049409341Z [2023-10-05 14:11:39 +0000] [9] [ERROR] Exception in worker process
2023-10-05T14:11:39.049440741Z Traceback (most recent call last):
2023-10-05T14:11:39.049447041Z   File "/venv/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-10-05T14:11:39.049452042Z     worker.init_process()
2023-10-05T14:11:39.049456442Z   File "/venv/lib/python3.11/site-packages/uvicorn/workers.py", line 66, in init_process
2023-10-05T14:11:39.049461042Z     super(UvicornWorker, self).init_process()
2023-10-05T14:11:39.049478642Z   File "/venv/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-10-05T14:11:39.049499143Z     self.load_wsgi()
2023-10-05T14:11:39.049504043Z   File "/venv/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-10-05T14:11:39.049508543Z     self.wsgi = self.app.wsgi()
2023-10-05T14:11:39.049512943Z                 ^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049517443Z   File "/venv/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-10-05T14:11:39.049521943Z     self.callable = self.load()
2023-10-05T14:11:39.049526343Z                     ^^^^^^^^^^^
2023-10-05T14:11:39.049530644Z   File "/venv/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-10-05T14:11:39.049535344Z     return self.load_wsgiapp()
2023-10-05T14:11:39.049539744Z            ^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049558944Z   File "/venv/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-10-05T14:11:39.049563344Z     return util.import_app(self.app_uri)
2023-10-05T14:11:39.049567445Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049571845Z   File "/venv/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-10-05T14:11:39.049576445Z     mod = importlib.import_module(module)
2023-10-05T14:11:39.049580845Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049586145Z   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-10-05T14:11:39.049591045Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-10-05T14:11:39.049595245Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049599745Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-10-05T14:11:39.049604445Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-10-05T14:11:39.049608846Z   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2023-10-05T14:11:39.049613246Z   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2023-10-05T14:11:39.049618146Z   File "/venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
2023-10-05T14:11:39.049622646Z     self.loader.exec_module(module)
2023-10-05T14:11:39.049640546Z   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
2023-10-05T14:11:39.049644846Z   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2023-10-05T14:11:39.049649347Z   File "/usr/src/app/copilot/main.py", line 13, in <module>
2023-10-05T14:11:39.049653847Z     from copilot.endpoints import auth, chat, document, user, zpages
2023-10-05T14:11:39.049658047Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-10-05T14:11:39.049662547Z   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2023-10-05T14:11:39.049667147Z   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2023-10-05T14:11:39.049672247Z   File "/venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
2023-10-05T14:11:39.049676747Z     self.loader.exec_module(module)
2023-10-05T14:11:39.049681047Z   File "/usr/src/app/copilot/endpoints/chat.py", line 10, in <module>
2023-10-05T14:11:39.049685548Z     from langchain.schema import AIMessage, messages_to_dict
2023-10-05T14:11:39.049690148Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-10-05T14:11:39.049694448Z   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2023-10-05T14:11:39.049699048Z   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2023-10-05T14:11:39.049723148Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/importer.py", line 177, in _exec_module
2023-10-05T14:11:39.049727849Z     notify_module_loaded(module)
2023-10-05T14:11:39.049732149Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/decorators.py", line 470, in _synchronized
2023-10-05T14:11:39.049736949Z     return wrapped(*args, **kwargs)
2023-10-05T14:11:39.049741849Z            ^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049746249Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/importer.py", line 136, in notify_module_loaded
2023-10-05T14:11:39.049751449Z     hook(module)
2023-10-05T14:11:39.049755749Z   File "/venv/lib/python3.11/site-packages/ddtrace/_monkey.py", line 167, in on_import
2023-10-05T14:11:39.049759849Z     imported_module.patch()
2023-10-05T14:11:39.049763949Z   File "/venv/lib/python3.11/site-packages/ddtrace/contrib/langchain/patch.py", line 759, in patch
2023-10-05T14:11:39.049768050Z     wrap("langchain", "llms.base.BaseLLM.generate", traced_llm_generate(langchain))
2023-10-05T14:11:39.049772450Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/wrappers.py", line 882, in wrap_function_wrapper
2023-10-05T14:11:39.049776950Z     return wrap_object(module, name, FunctionWrapper, (wrapper,))
2023-10-05T14:11:39.049797350Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049801950Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/wrappers.py", line 822, in wrap_object
2023-10-05T14:11:39.049806351Z     (parent, attribute, original) = resolve_path(module, name)
2023-10-05T14:11:39.049810551Z                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049827951Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/wrappers.py", line 808, in resolve_path
2023-10-05T14:11:39.049832651Z     original = lookup_attribute(parent, attribute)
2023-10-05T14:11:39.049837051Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049841551Z   File "/venv/lib/python3.11/site-packages/ddtrace/vendor/wrapt/wrappers.py", line 806, in lookup_attribute
2023-10-05T14:11:39.049845752Z     return getattr(parent, attribute)
2023-10-05T14:11:39.049849952Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-10-05T14:11:39.049854452Z   File "/venv/lib/python3.11/site-packages/langchain/__init__.py", line 328, in __getattr__
2023-10-05T14:11:39.049858552Z     raise AttributeError(f"Could not find: {name}")
2023-10-05T14:11:39.049878352Z AttributeError: Could not find: llms
```

--- a/ddtrace/contrib/langchain/constants.py
+++ b/ddtrace/contrib/langchain/constants.py
@@ -27,7 +27,7 @@ text_embedding_models = (
     "EmbaasEmbeddings",
 )
 
-vectorstores = (
+vectorstore_classes = (
     "AzureSearch",
     "Redis",
     "ElasticVectorSearch",

--- a/ddtrace/contrib/langchain/patch.py
+++ b/ddtrace/contrib/langchain/patch.py
@@ -20,7 +20,7 @@ from ddtrace.contrib.langchain.constants import PROVIDER
 from ddtrace.contrib.langchain.constants import TOTAL_COST
 from ddtrace.contrib.langchain.constants import TYPE
 from ddtrace.contrib.langchain.constants import text_embedding_models
-from ddtrace.contrib.langchain.constants import vectorstores
+from ddtrace.contrib.langchain.constants import vectorstore_classes
 from ddtrace.contrib.trace_utils import unwrap
 from ddtrace.contrib.trace_utils import with_traced_module
 from ddtrace.contrib.trace_utils import wrap
@@ -756,6 +756,14 @@ def patch():
             )
         integration.start_log_writer()
 
+    # Langchain doesn't allow wrapping directly from root, so we have to import the base classes first before wrapping.
+    # ref: https://github.com/DataDog/dd-trace-py/issues/7123
+    from langchain import embeddings  # noqa
+    from langchain import vectorstores  # noqa
+    from langchain.chains.base import Chain  # noqa
+    from langchain.chat_models.base import BaseChatModel  # noqa
+    from langchain.llms.base import BaseLLM  # noqa
+
     wrap("langchain", "llms.base.BaseLLM.generate", traced_llm_generate(langchain))
     wrap("langchain", "llms.base.BaseLLM.agenerate", traced_llm_agenerate(langchain))
     wrap("langchain", "chat_models.base.BaseChatModel.generate", traced_chat_model_generate(langchain))
@@ -777,7 +785,7 @@ def patch():
                 wrap("langchain", "embeddings.%s.embed_documents" % text_embedding_model, traced_embedding(langchain))
                 # TODO: langchain >= 0.0.209 includes async embedding implementation (only for OpenAI)
     # We need to do the same with Vectorstores.
-    for vectorstore in vectorstores:
+    for vectorstore in vectorstore_classes:
         if hasattr(langchain.vectorstores, vectorstore):
             # Ensure not double patched, as some Embeddings interfaces are pointers to other Embeddings.
             if not isinstance(
@@ -809,7 +817,7 @@ def unpatch():
                 deep_getattr(langchain.embeddings, "%s.embed_documents" % text_embedding_model), wrapt.ObjectProxy
             ):
                 unwrap(getattr(langchain.embeddings, text_embedding_model), "embed_documents")
-    for vectorstore in vectorstores:
+    for vectorstore in vectorstore_classes:
         if hasattr(langchain.vectorstores, vectorstore):
             if isinstance(
                 deep_getattr(langchain.vectorstores, "%s.similarity_search" % vectorstore), wrapt.ObjectProxy

--- a/releasenotes/notes/langchain-fix-imports-dddd44cd581e6baa.yaml
+++ b/releasenotes/notes/langchain-fix-imports-dddd44cd581e6baa.yaml
@@ -1,0 +1,5 @@
+---
+fixes:
+  - |
+    langchain: This fix resolves an import error with patching langchain versions newer than 0.0.300 due to
+    langchain dropping support for wrapping or importing directly from root. 

--- a/tests/contrib/langchain/test_langchain_patch.py
+++ b/tests/contrib/langchain/test_langchain_patch.py
@@ -2,7 +2,7 @@ from ddtrace.contrib.langchain import get_version
 from ddtrace.contrib.langchain import patch
 from ddtrace.contrib.langchain import unpatch
 from ddtrace.contrib.langchain.constants import text_embedding_models
-from ddtrace.contrib.langchain.constants import vectorstores
+from ddtrace.contrib.langchain.constants import vectorstore_classes
 from tests.contrib.patch import PatchTestCase
 
 
@@ -24,7 +24,7 @@ class TestLangchainPatch(PatchTestCase.Base):
             if embedding_model:
                 self.assert_wrapped(embedding_model.embed_query)
                 self.assert_wrapped(embedding_model.embed_documents)
-        for vectorstore in vectorstores:
+        for vectorstore in vectorstore_classes:
             vectorstore_interface = getattr(langchain.vectorstores, vectorstore, None)
             if vectorstore_interface:
                 self.assert_wrapped(vectorstore_interface.similarity_search)
@@ -41,7 +41,7 @@ class TestLangchainPatch(PatchTestCase.Base):
             if embedding_model:
                 self.assert_not_wrapped(embedding_model.embed_query)
                 self.assert_not_wrapped(embedding_model.embed_documents)
-        for vectorstore in vectorstores:
+        for vectorstore in vectorstore_classes:
             vectorstore_interface = getattr(langchain.vectorstores, vectorstore, None)
             if vectorstore_interface:
                 self.assert_not_wrapped(vectorstore_interface.similarity_search)
@@ -58,7 +58,7 @@ class TestLangchainPatch(PatchTestCase.Base):
             if embedding_model:
                 self.assert_not_double_wrapped(embedding_model.embed_query)
                 self.assert_not_double_wrapped(embedding_model.embed_documents)
-        for vectorstore in vectorstores:
+        for vectorstore in vectorstore_classes:
             vectorstore_interface = getattr(langchain.vectorstores, vectorstore, None)
             if vectorstore_interface:
                 self.assert_not_double_wrapped(vectorstore_interface.similarity_search)

